src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDynamicIndexerAccessExpression.cs

```diff
@@ -0,0 +1,401 @@
+                //     public int this[int i, ref dynamic ch] => 0;
+                Diagnostic(ErrorCode.ERR_IllegalRefParam, "ref").WithLocation(9, 28)
+            };
+
+            VerifyOperationTreeAndDiagnosticsForTest<ElementAccessExpressionSyntax>(source, expectedOperationTree, expectedDiagnostics);
+        }
+
+        [CompilerTrait(CompilerFeature.IOperation)]
+        [Fact]
+        public void DynamicIndexerAccessExpression_WithDynamicReceiver()
+        {
+            string source = @"
+class C
+{
+    void M(dynamic d, int i)
+    {
+        var x = /*<bind>*/d[i]/*</bind>*/;
+    }
+
+    public int this[int i] => 0;
+}
+";
+            string expectedOperationTree = @"
+IDynamicIndexerAccessExpression (OperationKind.DynamicIndexerAccessExpression, Type: dynamic) (Syntax: 'd[i]')
+  Expression: IParameterReferenceExpression: d (OperationKind.ParameterReferenceExpression, Type: dynamic) (Syntax: 'd')
+  ApplicableSymbols(0)
+  Arguments(1):
+      IParameterReferenceExpression: i (OperationKind.ParameterReferenceExpression, Type: System.Int32) (Syntax: 'i')
+  ArgumentNames(0)
+  ArgumentRefKinds(0)
+";
+            var expectedDiagnostics = DiagnosticDescription.None;
+
+            VerifyOperationTreeAndDiagnosticsForTest<ElementAccessExpressionSyntax>(source, expectedOperationTree, expectedDiagnostics);
+        }
+
+        [CompilerTrait(CompilerFeature.IOperation)]
+        [Fact]
+        public void DynamicIndexerAccessExpression_WithDynamicMemberReceiver()
+        {
+            string source = @"
+class C
+{
+    void M(dynamic c, int i)
+    {
+        var x = /*<bind>*/c.M2[i]/*</bind>*/;
+    }
+
+    public int this[int i] => 0;
+}
+";
+            string expectedOperationTree = @"
+IDynamicIndexerAccessExpression (OperationKind.DynamicIndexerAccessExpression, Type: dynamic) (Syntax: 'c.M2[i]')
+  Expression: IDynamicMemberReferenceExpression (Member Name: ""M2"", Containing Type: null) (OperationKind.DynamicMemberReferenceExpression, Type: dynamic) (Syntax: 'c.M2')
+      Type Arguments(0)
+      Instance Receiver: IParameterReferenceExpression: c (OperationKind.ParameterReferenceExpression, Type: dynamic) (Syntax: 'c')
+  ApplicableSymbols(0)
+  Arguments(1):
+      IParameterReferenceExpression: i (OperationKind.ParameterReferenceExpression, Type: System.Int32) (Syntax: 'i')
+  ArgumentNames(0)
+  ArgumentRefKinds(0)
+";
+            var expectedDiagnostics = DiagnosticDescription.None;
+
+            VerifyOperationTreeAndDiagnosticsForTest<ElementAccessExpressionSyntax>(source, expectedOperationTree, expectedDiagnostics);
+        }
+
+        [CompilerTrait(CompilerFeature.IOperation)]
+        [Fact]
+        public void DynamicIndexerAccessExpression_WithDynamicTypedMemberReceiver()
+        {
+            string source = @"
+class C
+{
+    dynamic M2 = null;
+
+    void M(dynamic c, int i)
+    {
+        var x = /*<bind>*/c.M2[i]/*</bind>*/;
```

Comment:
>c.M2[i] [](start = 26, length = 7)

I think it will also be useful to test indexing directly of off dynamic instance, i.e. ```c[i]```.

Reply:
Covered by DynamicIndexerAccessExpression_WithDynamicReceiver

---
In reply to: [137890762](https://github.com/dotnet/roslyn/pull/21711#discussion_r137890762) [](ancestors = 137890762)